[18.0][MIG] web_form_banner: Migration to 18.0

AungKoKoLin1997 opened the pull request:
Based on https://github.com/OCA/web/pull/3305

The module adds configurable banners for backend form views. Define rules per model
(and optionally per view) to show context-aware alerts with a chosen severity (info/warning/danger).

Messages can be plain text with ${placeholders} or fully custom HTML; visibility,
severity, and values are computed server-side via a safe Python expression.

Banners are injected just before or after a target node (default: //sheet) and refresh
on form load/save/reload.

@qrtl QT5747

yostashiro committed:
[ADD] web_form_banner

AungKoKoLin1997 committed:
[IMP] web_form_banner: black, isort, prettier

AungKoKoLin1997 committed:
[MIG] web_form_banner: Migration to 15.0

Refactor JS:
- Switch to the ESM style
- Change var to const/let
- Remove unused parts (diff, el.dataset.wfbTriggerFields)

Miscellaneous improvements:
- list view optional
- Switch archive box icon to banner

Co-authored-by: Yoshi Tashiro <tashiro@quartile.co>

AungKoKoLin1997 committed:
[IMP] web_form_banner: pre-commit auto fixes

AungKoKoLin1997 committed:
[MIG] web_form_banner: Migration to 16.0

yostashiro committed:
[IMP] web_form_banner: Partially support in-group target XPath

- Add (partial) support for elements inside groups to be a target
  XPath without distorting the presentation
- Remove obsolete bits

AungKoKoLin1997 committed:
[IMP] web_form_banner: pre-commit auto fixes

AungKoKoLin1997 committed:
[MIG] web_form_banner: Migration to 17.0

AungKoKoLin1997 committed:
[IMP] web_form_banner: pre-commit auto fixes

AungKoKoLin1997 committed:
[MIG] web_form_banner: Migration to 18.0

yostashiro commented on web_form_banner/demo/web_form_banner_rule_demo.xml:
```diff
@@ -0,0 +1,64 @@
+            name="trigger_field_ids"
+            eval="[(6, 0, [ref('base.field_res_partner__category_id')])]"
+        />
+        <field name="severity">warning</field>
```
```suggestion
        <field name="severity">danger</field>
```

yostashiro commented on web_form_banner/demo/web_form_banner_rule_demo.xml:
```diff
@@ -0,0 +1,64 @@
+            eval="[(6, 0, [ref('base.field_res_partner__category_id')])]"
+        />
+        <field name="severity">warning</field>
+        <field name="target_xpath">//sheet</field>
```
```suggestion
        <field name="target_xpath">//field[@name='category_id']</field>
```

yostashiro commented on web_form_banner/demo/web_form_banner_rule_demo.xml:
```diff
@@ -0,0 +1,64 @@
+        <field name="severity">warning</field>
+        <field name="target_xpath">//sheet</field>
+        <field name="position">before</field>
+        <field name="message">This partner is missing a tag!</field>
```
```suggestion
        <field name="message">Tag is missing!</field>
```

yostashiro reviewed (commented):
Just to demonstrate the possibility of showing a message inside a group.

bealdav reviewed (approved):
Tested functionnally.
really thanks

AungKoKoLin1997 committed:
[IMP] web_form_banner: imp readme

bealdav commented on web_form_banner/views/web_form_banner_rule_views.xml:
```diff
@@ -0,0 +1,208 @@
+                        <h1>
+                            <field
+                                name="name"
+                                placeholder="e.g. Warning on dangeours customers"
```
```suggestion
                                placeholder="e.g. Warning on dangerous customers"
```

AungKoKoLin1997 replied:
Fixed!

yostashiro commented on web_form_banner/models/web_form_banner_rule.py:
```diff
@@ -98,7 +98,9 @@ class WebFormBannerRule(models.Model):
     model_id = fields.Many2one("ir.model", ondelete="cascade", required=True)
-    model_name = fields.Char(related="model_id.model", store=True, readonly=True)
+    model_name = fields.Char(
+        related="model_id.model", string="Model Name", store=True, readonly=True
```
Why do we need to add `string`?

AungKoKoLin1997 replied:
I found the default label of this field is `Model` and same with label of `model_id`.
So, that causes the pre-commit issue and added `string` to fix.

yostashiro reviewed (approved):
Code review. LGTM.

yostashiro reviewed (changes requested):
I just noticed an issue. The banner doesn't seem to reload when records are switched with previous/next button.

<img width="194" height="84" alt="image" src="https://github.com/user-attachments/assets/ba470fcb-d709-4540-b3a9-0156ddf75bba" />

This issue seems to be present in the current 17.0 and 18.0 PRs. It's not the case with 16.0 (the banner reloads fine).

@AungKoKoLin1997 Can you please look into this?

AungKoKoLin1997 committed:
[FIX] web_form_banner

- Fix not updating banner when pager is update or duplicate or create
- Fix presentation of help in web form banner rule

yostashiro commented on web_form_banner/views/web_form_banner_rule_views.xml:
```diff
@@ -0,0 +1,204 @@
+                                    Should be used instead of <code
+                                >record</code> when your rule is triggered dynamically by an
+                                    update to a trigger field. It doesn't include any values from complex fields
+                                    (x2many/reference, etc).
```
```suggestion
                                    (one2many/reference, etc).
```

yostashiro reviewed (approved):
Code review and functional test. LGTM.

kanda999 reviewed (approved):
Functioanl review: it works as expected

sebastienbeau commented:
/ocabot merge nobump
